sdks/unity/AgonesSdk.cs

```diff
@@ -0,0 +1,239 @@
+﻿// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using UnityEngine;
+using UnityEngine.Networking;
+
+namespace Agones
+{
+    /// <summary>
+    /// Agones SDK for Unity.
+    /// </summary>
+    public class AgonesSdk : MonoBehaviour
+    {
+        [Range(1, 5)]
+        public float healthIntervalSecond = 5.0f;
+        public bool healthEnabled = true;
+        public bool logEnabled = false;
+
+        private const string sidecarAddress = "http://localhost:59358";
+        private readonly CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+
+        private float CurrentHealthTime { get; set; } = 0;
+
+        private struct KeyValueMessage
+        {
+            public string key;
+            public string value;
+            public KeyValueMessage(string k, string v) => (key, value) = (k, v);
+        }
+
+        #region Unity Methods
+        // Use this for initialization
+        void Start()
+        {
+            CurrentHealthTime = 0;
+        }
+
+        // Update is called once per frame
+        void Update()
+        {
+            if (!healthEnabled) { return; }
+
+            CurrentHealthTime += Time.unscaledDeltaTime;

```

Comment:
I think this health check would be more appropriate as a Coroutine as opposed to being checked every Update. This would cut out `CurrentHealthTime` entirely and avoid running code on every Update cycle.

In this example, the `Health()` method would also need to also be an async function that returns `true` if the health was successful. That part is recommended, but not required.

```csharp
private void Create() 
{
  if (healthEnabled) {
    StartCoroutine(HealthCheck());
  }
}

private IEnumerator HealthCheck()
{
  while (true) 
  {
    yield return new WaitForSeconds(healthIntervalSecond);
    if (await Health() == false) 
    {
      Debug.Log("Health check failed.");
    }
  }
}
```

Reply:
`await` can not be used in a not async function, can I use `async`?
like the below code.
```csharp
private async void HealthCheckAsync()
{
    while (healthEnabled)
    {
        await Task.Delay(TimeSpan.FromSeconds(healthIntervalSecond));

        try
        {
            await SendRequestAsync("/health", "{}");
        }
        catch (ObjectDisposedException)
        {
        }
    }
}
```

And if `LogEnabled` checked, logs will be output when `Health()` fails.
Should I still check the return value and output logs?

Reply:
Unity supports async and await from at least 2017 and beyond, so you can indeed use async function. The only thing to verify (because I'm not sure off hand) is how to properly trigger the function on a separate thread instead of holding up the main thread. It might be as simple as calling it from start as normal (without the `await` keyword).